File: ProcessTRX/filterTRX.py
#!/usr/bin/python
'''
This script first takes a filter and uses it prune down a large TRX file so that it only contains
  the aircraft seen in the filter.  Then, on the reduced TRX, it attempts to keep only the first
  track time seen for each aircraft.  This latter step is necessary because FACET has issues with
  simulations and track updates, so just remove the track updates and let it simulate the flight
  based on the first state vector.

  Author: Thomas J Colvin
'''

import sys
import os.path

print 'Number of arguments:', len(sys.argv), 'arguments.'
print 'Argument List:', str(sys.argv)

if len(sys.argv) != 4:
    print "   Error: need exactly three arguments after file name"
    print "          fileToFilterOn fileToPrune outputFile"
    sys.exit()

# Unpack inputs
fileToFilterOn  = sys.argv[1]
fileToPrune     = sys.argv[2]
outputFile      = sys.argv[3]

# Check that inputs exist
if not os.path.isfile(fileToFilterOn):
    print "   Error: {0} does not exist".format(fileToFilterOn)
    sys.exit()
elif not os.path.isfile(fileToPrune):
    print "   Error: {0} does not exist".format(fileToPrune)
    sys.exit()


'''
Okay, the inputs check out.  Now run the actual script.
'''

from TRX_functions import readTRXIntoDictByCallsign
from TRX_functions import pruneTRXFromList

# fileToFilterOn  = 'TRX_Files/CapeShapeFilteredAC02'
# fileToPrune     = 'TRX_Files/TRX_20050824_FastForward_OnTheMinute'
# outputFile      = 'TRX_20050824_FastForward_OnTheMinute_Filtered'

# fileToFilterOn  = 'Kevin2018/ColoradoAircraftFilter'
# fileToPrune     = 'Kevin2018/TRX_TAF_PlannedTestFull'
# outputFile      = 'TempTRX'

tempOutFileName = "OutputTRX/TempTRX"

# Just load the entire TRX, since it's been filtered, it should be relatively small
TRX_Storage = readTRXIntoDictByCallsign(fileToFilterOn)

# All the keys are the flights you want.  The dep airport is appended to each for (quasi) uniqueness; remove that
# First just whittle it down to only the flights you want
sortedKeys      = sorted(TRX_Storage.keys())
flightsToKeep   = [curFlight.split('_')[0] for curFlight in sortedKeys]
filterDepAirports   = False
filterTimes         = False
bigFilterKeys       = []
filterDict = {'callsigns' : flightsToKeep, 'sectorNames' : [], 'filterDepAirports' : filterDepAirports,
              'filterTimes' : filterTimes, 'bigFilterKeys' : bigFilterKeys}
# pruneTRXFromList(fileToPrune, filterDict, outputFile)
pruneTRXFromList(fileToPrune, filterDict, tempOutFileName)


'''
# Now that you've eliminated all the extra flights, make sure things are clean by
#   * Pruning based on departure airport
#   * Only keeping the first track time (same as removing updates)
'''

fileToPrune     = tempOutFileName
# outputFile      = 'TRX_TAF_PlannedTestFull_Colorado'
TRX_Storage = readTRXIntoDictByCallsign(fileToPrune)

sortedKeys      = sorted(TRX_Storage.keys())
flightsToKeep   = [curFlight.split('_')[0] for curFlight in sortedKeys]
depAirports     = [curFlight.split('_')[1] for curFlight in sortedKeys]
firstTimes      = [set([TRX_Storage[curFlight][0][0]]) for curFlight in sortedKeys]

import numpy as np

newFlightThresh = 10 * 60   # Minutes * seconds
for (ix, curFlight) in enumerate(sortedKeys):
    curTimes = np.array(TRX_Storage[curFlight], np.int64)[:,0]
    threshIX = np.diff(curTimes) > newFlightThresh

    if sum(threshIX) > 0:
        # This means that there are multiple flights with this callsign_depAirport present.
        newFirstTimeIX = np.where(threshIX)[0] + 1
        newTimes = set(curTimes[newFirstTimeIX])

        firstTimes[ix] = firstTimes[ix].union(newTimes)

# TODO: Prepend proper on-the-minute track time
# TODO: Simulating the output of this script, I'm seeing lots of flights that don't get anywhere NEAR Denver. Why?

filterDepAirports   = True
filterTimes         = True
# bigFilterKeys = ['{0}_{1}_{2}'.format(callsign, dep, time) for (callsign, dep, time) in zip(flightsToKeep, depAirports, firstTimes)]
bigFilterKeys = ['{0}_{1}_{2}'.format(callsign, dep, curTime)
                 for (callsign, dep, timeSet) in zip(flightsToKeep, depAirports, firstTimes)
                 for curTime in timeSet]

### Now filter the main TRX using flightsToKeep
# First package the info up so we can jam it into an existing function
filterDict = {'callsigns' : flightsToKeep, 'sectorNames' : [], 'filterDepAirports' : filterDepAirports,
              'filterTimes' : filterTimes, 'bigFilterKeys' : bigFilterKeys}
pruneTRXFromList(fileToPrune, filterDict, outputFile)

# key = pruneTRXFromList(fileToPrune, filterDict, outputFile)

File: ProcessTRX/filterTRX.jl
# '''
# Take a list of CallSigns and remove every track update that comes from an AC with that callsign
# '''


function pruneTRXFromList(inFileName::String, filterDict::Dict, outFileName::String, deltaSeconds = 0)
  # Starts by eliminating every single aircraft (shouldPrint = False), but if an AC matches a filter,
  #   then it will be allowed through.

  print("Pruning for " * outFileName)
  sortedKeys = sort(collect(keys(filterDict)))
  plainCallsigns = [split(curFlight, '_')[1] for curFlight in sortedKeys]

  # Keys are TRACK_TIMES, each value will be the strings (concatenated together) that are to be printed out
  TRX_Storage = Dict()

  printOkay = true
  if length(outFileName) == 0
    printOkay = false
  end

  inputFile = open(inFileName, "r")

  # This will hold the information as we run through the lines until it's time to save it to the dictionary
  curTime = -1
  curFPLine = ""
  curTrackLine = ""
  shouldPrint = true
  curCallSign = []    # Scope this here

  for line in eachline(inputFile)
    key = split(strip(line))

    if length(key) > 0
      #println(key)
      #println(length(key))

      if key[1] == "TRACK_TIME"
        # Indicates progress

        curTimeInt = int(key[2]) + deltaSeconds
        curTime = string(curTimeInt)

      elseif key[1] == "TRACK"
        curCallSign = key[2]

#         shouldPrint = (curCallSign in plainCallsigns)
#         shouldPrint = true
        curTrackLine = line

      elseif key[1] == "FP_ROUTE"
        curFPLine = line

        # At this point we need to check for the departure Airport
#         if shouldPrint
          #               depAirport = key[1].split('.')[0]
          depAirport = split(key[2],".")[1]
          sign_dep = "$(curCallSign)_$depAirport"

          if !(sign_dep in sortedKeys)
            shouldPrint = false
          elseif !(curTimeInt in filterDict[sign_dep])
            shouldPrint = false
          end
#         end


        # At this point, we will be done with the specific aircraf that we're looking at so save it if it's good
        if shouldPrint
          # Create this time record if necessary
          if !haskey(TRX_Storage, curTime)
            TRX_Storage[curTime] = ""
          end

          # Add it to the queue for printing
          TRX_Storage[curTime] = TRX_Storage[curTime] * curTrackLine
          TRX_Storage[curTime] = TRX_Storage[curTime] * curFPLine * "\n"

        else
          # Reset
          shouldPrint = true
        end

      elseif key[1][1] == "#"
        # print 'COMMENT'
      else
        print("pruneTRXFromList: NewLine! This should not have been hit!\n")
        print(key)
        error()
      end
    end
  end

  close(inputFile) # It's likely that the output file will be the same file, so make sure to close this first.

  if printOkay
    # # Now write everything to file
    outFile = open(outFileName,"w")

    for key in sort(collect(keys(TRX_Storage)))
      if length(TRX_Storage[key]) > 0
        write(outFile, "TRACK_TIME " * key * "\n")
        write(outFile, TRX_Storage[key])
      end
    end

    close(outFile)
  end
end


cd(dirname(@__FILE__))   # Set pwd to location of this file
ENV["PYTHONPATH"] = pwd()   # Tell python where to look for modules
using PyCall
@pyimport TRX_functions
readTRXIntoDictByCallsign = TRX_functions.readTRXIntoDictByCallsign


filterPath = "/Users/tcolvin1/Desktop/aiaaSpace/FilteredTRX/"
rawPath    = "/Users/tcolvin1/Desktop/aiaaSpace/RawTRX/"

fileToFilterOn = filterPath * "TRX_FrontRange450_TRX_DW2015013119201337.trx"
# fileToPrune    = rawPath * "TRX_DW2015013119201337.trx"
fileToPrune    = rawPath * "testRaw1337.trx"


# from TRX_functions import readTRXIntoDictByCallsign
# from TRX_functions import pruneTRXFromList

tempOutFileName = "TempTRX"   # This is good for debugging
# tempOutFileName = outputFile  # This should be okay if everything is working...

# Just load the entire TRX, since it's been filtered, it should be relatively small
TRX_Storage = readTRXIntoDictByCallsign(fileToFilterOn)
sortedKeys = sort(collect(keys(TRX_Storage)))


curTimes = TRX_Storage[sortedKeys[1]][:,1]
jumpIXs = find(diff(curTimes)/60 .> 5)
keepIXs = unique(vcat([1], jumpIXs+1, [length(curTimes)]))
curTimes[keepIXs]

(936000/388. * 0.75)/60   #minutes.  I don't think it took 30 minutes in python!

# jumpIXs are the LAST times before the jump.
# Have to add 1 to index to get the first time of the next flight.

# So will want to keep times for all of the jumps, the first time seen and the last time seen
filterDict = Dict()

for curCallSign in sortedKeys
  curTimes = TRX_Storage[curCallSign][:,1]
  jumpIXs = find(diff(curTimes)/60 .> 5)
  keepIXs = unique(vcat([1], jumpIXs+1, [length(curTimes)]))
  filterDict[curCallSign] = curTimes[keepIXs]

end

filterDict

@time pruneTRXFromList(fileToPrune, filterDict, tempOutFileName)
@profile pruneTRXFromList(fileToPrune, filterDict, tempOutFileName)
Profile.print()






